mcs/class/corlib/Test/System.Reflection.Emit/MethodBuilderTest2.cs

```diff
@@ -0,0 +1,154 @@
+						return false;
+					var params1 = m.GetParameters ();
+					var params2 = methodBuilder.GetParameters ();
+					if (params1.Length != params2.Length)
+						return false;
+					for (var i = 0; i < params1.Length; i++)
+						if (params1 [i].ParameterType.FullName != params2 [i].ParameterType.FullName)
+							return false;
+
+					return true;
+				});
+
+				object inst = Activator.CreateInstance (type, new object [0]);
+
+				return method.Invoke (inst, parameters);
+			}
+
+			[Test]
+			public void CallMethodRef ()
+			{
+				var expected = "value";
+
+				var moduleBuilder = CreateModuleBuilder ();
+				var typeBuilder = moduleBuilder.DefineType ("NewType");
+
+				var methodBuilder1 = typeBuilder.DefineMethod ("NewMethod1",
+					MethodAttributes.Public | MethodAttributes.Static,
+					typeof (string),
+					Type.EmptyTypes);
+
+				var gen1 = methodBuilder1.GetILGenerator ();
+				gen1.Emit (OpCodes.Ldstr, expected);
+				gen1.Emit (OpCodes.Ret);
+
+				var methodBuilder2 = typeBuilder.DefineMethod ("NewMethod2",
+				MethodAttributes.Public | MethodAttributes.Static,
+					typeof (string),
+					Type.EmptyTypes);
+
+				var ilStream = new MemoryStream ();
+				var ilWriter = new BinaryWriter (ilStream);
+				ilWriter.Write ((byte) 0x28); // call
+				ilWriter.Write ((int)  moduleBuilder.GetMethodToken (methodBuilder1).Token);
+				ilWriter.Write ((byte) 0x2A); // ret
+
+				SetIL (methodBuilder2, ilStream);
+
+				var type = typeBuilder.CreateType ();
+
+				Assert.AreEqual (expected, Invoke (type, methodBuilder2));
+			}
+
+			[Test]
+			public void CallMethodDef ()
+			{
+				var expected = "value";
+
+				var moduleBuilder1 = CreateModuleBuilder ();
+				var typeBuilder1 = moduleBuilder1.DefineType ("NewType1", TypeAttributes.Public);
+
+				var methodBuilder1 = typeBuilder1.DefineMethod ("NewMethod1",
+				 MethodAttributes.Public | MethodAttributes.Static,
+					typeof (string),
+					Type.EmptyTypes);
+
+				var gen1 = methodBuilder1.GetILGenerator ();
+				gen1.Emit (OpCodes.Ldstr, expected);
+				gen1.Emit (OpCodes.Ret);
+
+				typeBuilder1.CreateType ();
+
+				var moduleBuilder2 = CreateModuleBuilder ();
+				var typeBuilder2 = moduleBuilder2.DefineType ("NewType2");
+
+				var methodBuilder2 = typeBuilder2.DefineMethod ("NewMethod2",
+					MethodAttributes.Public | MethodAttributes.Static,
+					typeof (string),
+					Type.EmptyTypes);
+
+				var ilStream = new MemoryStream ();
+				var ilWriter = new BinaryWriter (ilStream);
+				ilWriter.Write ((byte) 0x28); // call
+				ilWriter.Write ((int) moduleBuilder2.GetMethodToken (methodBuilder1).Token);
+				ilWriter.Write ((byte) 0x2A); // ret
+
+				SetIL (methodBuilder2, ilStream);
+
+				var type = typeBuilder2.CreateType ();
+
+				Assert.AreEqual (expected, Invoke (type, methodBuilder2));
+			}
+		}
+
+		/*
+		 * Tests MethodBuilder.CreateMethodBody
+		 */
+		[TestFixture]
+		public class MethodBuilderTest2_CreateMethodBody : MethodBuilderTest2
```

Comment:
The two additional classes are taking advantage of inheritance to test CreateMethodBody and SetMethodBody with the same tests implemented in the base class.